esp8266/modesp.c

```diff
@@ -35,6 +35,492 @@
+    char ip_str[16];
+    mp_uint_t ip_len = snprintf(ip_str, 16, "%u.%u.%u.%u", ip[0], ip[1], ip[2], ip[3]);
+    return mp_obj_new_str(ip_str, ip_len, false);
+}
+
+// Takes an array with a raw IP address, and a port, and returns a net-address
+// tuple such as ('192.168.0.1', 8080).
+STATIC mp_obj_t esp_espconn_format_inet_addr(uint8_t *ip, mp_uint_t port) {
+    mp_obj_t tuple[2] = {
+        tuple[0] = esp_espconn_format_ipv4_addr(ip),
+        tuple[1] = mp_obj_new_int(port),
+    };
+    return mp_obj_new_tuple(2, tuple);
+}
+
+STATIC const mp_obj_type_t esp_espconn_type;
+
+typedef struct _esp_espconn_obj_t {
+    mp_obj_base_t base;
+    struct espconn *espconn;
+
+    mp_obj_t cb_connect;
+    mp_obj_t cb_recv;
+    mp_obj_t cb_disconnect;
+
+    uint8_t *recvbuf;
+    mp_uint_t recvbuf_len;
+
+    bool server;
+
+    mp_obj_list_t *connlist;
+} esp_espconn_obj_t;
+
+// Due to the onconnect callback not being able to recognize the parent esp_espconn,
+// we can have only one esp_espconn listening at a time
+// This should be solvable by some PIC hacking
+STATIC esp_espconn_obj_t *esp_espconn_listening;
+
+STATIC mp_obj_t esp_espconn_make_new_base() {
+    esp_espconn_obj_t *s = m_new_obj_with_finaliser(esp_espconn_obj_t);
+    s->recvbuf = NULL;
+    s->base.type = (mp_obj_t)&esp_espconn_type;
+    s->cb_connect = mp_const_none;
+    s->cb_recv = mp_const_none;
+    s->cb_disconnect = mp_const_none;
+    s->server = false;
+    return s;
+}
+
+// constructor esp_espconn(family=AF_INET, type=SOCK_STREAM, proto=IPPROTO_TCP, fileno=None)
+// Arguments ignored as we do not support UDP (yet)
+STATIC mp_obj_t esp_espconn_make_new(mp_obj_t type_in, mp_uint_t n_args,
+    mp_uint_t n_kw, const mp_obj_t *args) {
+    mp_arg_check_num(n_args, n_kw, 0, 4, false);
+
+    esp_espconn_obj_t *s = esp_espconn_make_new_base();
+    s->espconn = gc_alloc(sizeof(struct espconn), false);
+    espconn_create(s->espconn);
+
+    s->espconn->reverse = s;
+    // TODO: UDP Support
+    s->espconn->type = ESPCONN_TCP;
+    s->espconn->state = ESPCONN_NONE;
+
+    s->espconn->proto.tcp = gc_alloc(sizeof(esp_tcp), false);
+
+    return s;
+}
+
+// method espconn.close()
+STATIC mp_obj_t esp_espconn_close(mp_obj_t self_in) {
+    esp_espconn_obj_t *s = self_in;
+
+    if (esp_espconn_listening == s)
+        esp_espconn_listening = NULL;
+
+    if (s->espconn->state != ESPCONN_NONE && s->espconn->state != ESPCONN_CLOSE) {
+        espconn_disconnect(s->espconn);
+    }
+
+    if (s->connlist != NULL) {
+        mp_obj_list_set_len(s->connlist, 0);
+    }
+
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(esp_espconn_close_obj, esp_espconn_close);
+
+// method socket.__del__()
+STATIC mp_obj_t esp_espconn___del__(mp_obj_t self_in) {
+    esp_espconn_obj_t *s = self_in;
+
+    esp_espconn_close(self_in);
+
+    if (s->server) {
+        espconn_delete(s->espconn);
+    } else {
+        gc_free(s->espconn->proto.tcp);
```

Comment:
Actually, you shouldn't free this memory explicitly.  If the GC has already marked the espconn object for reclamation then it already marked these buffers and freeing them explicitly may lead to undefined behaviour.